File: tools/py/bench.py
#!/usr/bin/env python3
import json, os, re, sys
from math import ceil

def die(msg):
    print(msg, file=sys.stderr)
    sys.exit(1)

def read_json(p):
    with open(p, "r", encoding="utf-8") as f:
        return json.load(f)

def list_files_recursive(inputs):
    out = []
    for p in inputs:
        if os.path.isdir(p):
            for root, _, files in os.walk(p):
                for fn in files:
                    out.append(os.path.join(root, fn))
        elif os.path.isfile(p):
            out.append(p)
    out.sort()
    return out

def canonicalize_to_llmd_lines(text: str, cfg: dict) -> list[str]:
    text = text.replace("\r\n", "\n")
    raw_lines = text.split("\n")

    lines = []
    if cfg.get("ignore_blocks", True):
        in_fence = False
        for ln in raw_lines:
            t = ln.strip()
            if t.startswith("```"):
                in_fence = not in_fence
                continue
            if in_fence:
                continue
            lines.append(ln)
    else:
        lines = raw_lines

    out = []
    current = None

    def emit(scope: str):
        nonlocal current
        if scope and scope != current:
            out.append("@" + scope)
            current = scope

    def norm_scope(s: str) -> str:
        return re.sub(r"\s+", "_", s.strip()).lower()

    def norm_key(s: str) -> str:
        s = re.sub(r"\s+", "_", s.strip().lower())
        s = re.sub(r"[^a-z0-9_]", "", s)
        return s

    for ln0 in lines:
        ln = ln0.strip()
        if not ln:
            continue

        m = re.match(r"^(#{1,6})\s+(.*)$", ln)
        if m:
            emit(norm_scope(m.group(2)))
            continue

        m = re.match(r"^[-*+]\s+(.*)$", ln) or re.match(r"^\d+\.\s+(.*)$", ln)
        if m:
            if not current:
                emit("root")
            out.append(">" + m.group(1).strip())
            continue

        m = re.match(r"^([A-Za-z0-9 _-]{2,64})\s*:\s*(.+)$", ln)
        if m:
            if not current:
                emit("root")
            k = norm_key(m.group(1))
            v = m.group(2).strip()
            out.append(":" + k + "=" + v if k else ">" + ln)
            continue

        if not current:
            emit("root")
        out.append(">" + ln)

    return out

def est_tokens_line(line: str) -> int:
    toks = [t for t in re.split(r"\s+", line.strip()) if t]
    return sum(ceil(len(t)/4) for t in toks)

def apply_dict_token_mode(lines: list[str], d: dict) -> list[str]:
    maps = d.get("maps", {})
    scope_map = maps.get("scope", {})
    key_map = maps.get("key", {})
    value_map = maps.get("value", {})
    text_map = maps.get("text", {})
    type_map = maps.get("type", {})

    out = []
    for line in lines:
        if line.startswith("@"):
            s = line[1:].strip().lower()
            out.append("@" + scope_map.get(s, s))
            continue

        if line.startswith(":"):
            parts = [p for p in re.split(r"\s+", line[1:].strip()) if p]
            new_parts = []
            for p in parts:
                if "=" not in p:
                    new_parts.append(p)
                    continue
                k, v = p.split("=", 1)
                k2 = key_map.get(k.lower(), k.lower())

                # Split enum chunks but keep delimiters
                chunks = re.split(r"([|,])", v)
                out_chunks = []
                for ch in chunks:
                    if ch in ("|", ","):
                        out_chunks.append(ch)
                        continue
                    low = ch.strip().lower()
                    if re.fullmatch(r"[a-z][a-z0-9._-]*", low) and low in value_map:
                        out_chunks.append(value_map[low])
                    else:
                        out_chunks.append(ch.strip())
                v2 = "".join(out_chunks)
                new_parts.append(f"{k2}={v2}")

            out.append(":" + " ".join(new_parts))
            continue

        if line.startswith(">"):
            toks = [t for t in re.split(r"\s+", line[1:].strip()) if t]
            nt = []
            for t in toks:
                low = t.lower()
                low = re.sub(r"^[^a-z0-9._-]+", "", low)
                low = re.sub(r"[^a-z0-9._-]+$", "", low)
                nt.append(text_map.get(low, t))
            out.append(">" + " ".join(nt))
            continue

        if line.startswith("::"):
            tp = line[2:].strip().lower()
            out.append("::" + type_map.get(tp, tp))
            continue

        out.append(line)

    return out

# CLI: python bench.py <config.json> <dict.json> <inputs...>
argv = sys.argv[1:]
if len(argv) < 3:
    die("Usage: python bench.py <config.json> <dict.json> <inputs...>")

cfg = read_json(argv[0])
d = read_json(argv[1])
files = list_files_recursive(argv[2:])
if not files:
    die("No input files found.")

before = 0
after = 0

for fp in files:
    with open(fp, "r", encoding="utf-8") as f:
        txt = f.read()
    ll = canonicalize_to_llmd_lines(txt, cfg)
    ll2 = apply_dict_token_mode(ll, d)
    before += sum(est_tokens_line(x) for x in ll)
    after += sum(est_tokens_line(x) for x in ll2)

saved = before - after
pct_red = (saved / before * 100) if before else 0.0
pct_size = (after / before * 100) if before else 0.0

print(f"Files: {len(files)}")
print(f"Est tokens BEFORE: {before}")
print(f"Est tokens AFTER : {after}")
print(f"Saved: {saved} ({pct_red:.1f}% reduction, final size {pct_size:.1f}%)")

File: tools/js/bench.js
#!/usr/bin/env node
import fs from "fs";
import path from "path";

function die(msg){ console.error(msg); process.exit(1); }
function readJSON(p){ return JSON.parse(fs.readFileSync(p,"utf8")); }

function listFilesRecursive(inputs) {
  const out = [];
  for (const p of inputs) {
    const st = fs.statSync(p);
    if (st.isDirectory()) {
      const entries = fs.readdirSync(p).map(e => path.join(p, e));
      out.push(...listFilesRecursive(entries));
    } else if (st.isFile()) out.push(p);
  }
  out.sort((a,b)=>a.localeCompare(b));
  return out;
}

function normalizeUnicodeNFKC(s){ return s.normalize("NFKC"); }

function canonicalizeToLLMDLines(text, cfg) {
  text = normalizeUnicodeNFKC(text).replace(/\r\n/g, "\n");
  const raw = text.split("\n");

  let lines = [];
  if (cfg.ignore_blocks) {
    let inFence = false;
    for (const ln of raw) {
      const t = ln.trim();
      if (t.startsWith("```")) { inFence = !inFence; continue; }
      if (inFence) continue;
      lines.push(ln);
    }
  } else lines = raw;

  const out = [];
  let cur = null;
  const emit = (s) => { if (s && s !== cur) { out.push("@"+s); cur = s; } };
  const normScope = s => s.trim().replace(/\s+/g,"_").toLowerCase();
  const normKey = s => s.trim().toLowerCase().replace(/\s+/g,"_").replace(/[^a-z0-9_]/g,"");

  for (const ln0 of lines) {
    const ln = ln0.trim();
    if (!ln) continue;

    const mh = ln.match(/^(#{1,6})\s+(.*)$/);
    if (mh) { emit(normScope(mh[2])); continue; }

    const mb = ln.match(/^[-*+]\s+(.*)$/) || ln.match(/^\d+\.\s+(.*)$/);
    if (mb) { if (!cur) emit("root"); out.push(">"+mb[1].trim()); continue; }

    const mkv = ln.match(/^([A-Za-z0-9 _-]{2,64})\s*:\s*(.+)$/);
    if (mkv) { if (!cur) emit("root"); const k = normKey(mkv[1]); out.push(k?(":"+k+"="+mkv[2].trim()):(">"+ln)); continue; }

    if (!cur) emit("root");
    out.push(">"+ln);
  }
  return out;
}

function estTokensLine(line) {
  // ignore blocks (none here), sum ceil(len(tok)/4) over whitespace toks
  const toks = line.split(/\s+/).filter(Boolean);
  let sum = 0;
  for (const t of toks) sum += Math.ceil(t.length / 4);
  return sum;
}

function applyDictTokenMode(lines, dict) {
  const maps = dict.maps || {};
  const scopeMap = maps.scope || {};
  const keyMap = maps.key || {};
  const valueMap = maps.value || {};
  const textMap = maps.text || {};
  const typeMap = maps.type || {};

  const out = [];
  for (const line of lines) {
    if (line.startsWith("@")) {
      const s = line.slice(1).trim().toLowerCase();
      out.push("@"+(scopeMap[s] ?? s));
      continue;
    }
    if (line.startsWith(":")) {
      const parts = line.slice(1).trim().split(/\s+/).filter(Boolean);
      const nParts = parts.map(p => {
        const i = p.indexOf("=");
        if (i <= 0) return p;
        const k = p.slice(0,i).toLowerCase();
        const v = p.slice(i+1);
        const nk = keyMap[k] ?? k;
        // enum split on | and , then remap eligible tokens
        const vv = v.split(/([|,])/g).map(chunk => {
          const c = chunk.trim();
          const low = c.toLowerCase();
          if (chunk === "|" || chunk === ",") return chunk;
          if (/^[a-z][a-z0-9._-]*$/.test(low) && valueMap[low]) return valueMap[low];
          return c;
        }).join("");
        return nk + "=" + vv;
      });
      out.push(":" + nParts.join(" "));
      continue;
    }
    if (line.startsWith(">")) {
      const toks = line.slice(1).trim().split(/\s+/).filter(Boolean);
      const nt = toks.map(t => {
        const low = t.toLowerCase().replace(/^[^a-z0-9._-]+/g,"").replace(/[^a-z0-9._-]+$/g,"");
        return textMap[low] ?? t;
      });
      out.push(">" + nt.join(" "));
      continue;
    }
    if (line.startsWith("::")) {
      const t = line.slice(2).trim().toLowerCase();
      out.push("::" + (typeMap[t] ?? t));
      continue;
    }
    out.push(line);
  }
  return out;
}

// CLI: node bench.js <config.json> <dict.json> <inputs...>
const argv = process.argv.slice(2);
if (argv.length < 3) die("Usage: node bench.js <config.json> <dict.json> <inputs...>");
const cfg = readJSON(argv[0]);
const dict = readJSON(argv[1]);
const files = listFilesRecursive(argv.slice(2));
if (!files.length) die("No input files found.");

let before = 0;
let after = 0;

for (const fp of files) {
  const txt = fs.readFileSync(fp, "utf8");
  const ll = canonicalizeToLLMDLines(txt, cfg);
  const ll2 = applyDictTokenMode(ll, dict);

  for (const ln of ll) before += estTokensLine(ln);
  for (const ln of ll2) after += estTokensLine(ln);
}

const saved = before - after;
const pctRed = before ? (saved / before) * 100 : 0;
const pctSize = before ? (after / before) * 100 : 0;

console.log(`Files: ${files.length}`);
console.log(`Est tokens BEFORE: ${before}`);
console.log(`Est tokens AFTER : ${after}`);
console.log(`Saved: ${saved} (${pctRed.toFixed(1)}% reduction, final size ${pctSize.toFixed(1)}%)`);